### PR TITLE
arch-riscv: Add support for riscv hardware probing syscall

### PR DESCRIPTION
**Pull request by 5265325**
This PR adds the support for riscv hardware probing syscall described in [this](https://docs.kernel.org/arch/riscv/hwprobe.html). The implementation logic refers to [linux kernel](https://github.com/torvalds/linux/blob/master/arch/riscv/kernel/sys_hwprobe.c) and [qemu](https://github.com/qemu/qemu/blob/master/linux-user/syscall.c). And passed the [RISC-V hwprobe exmaple](https://github.com/cyyself/hwprobe) test.

Hope to be merged. Thanks.

Change-Id: Iab714974f0551fc451e0d6846c75a7153809a308

### CODE REVIEW

File: src/arch/riscv/linux/linux.hh [hunk 1]
```diff
@@ -42,6 +42,86 @@ class RiscvLinux : public Linux
 {
   public:
     static const ByteOrder byteOrder = ByteOrder::little;
+
+    #define RISCV_HWPROBE_KEY_MVENDORID	0
```
Comment (powerjg) [inline src/arch/riscv/linux/linux.hh]: I think this would be better implemented as a `BitUnion`

See for instance the `types.hh` file.

File: src/arch/riscv/linux/se_workload.cc [hunk 2]
```diff
@@ -123,6 +125,179 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+    BufferArg buf(cpus, target_size);
+    buf.copyIn(SETranslatingPortProxy(tc));
+
+    for (i = 0 ; i < target_size; i++) {
+        unsigned bit = i * target_bits;
+        uint8_t val = *((uint8_t *)buf.bufferPtr() + i);
+
+        for (j = 0; j < target_bits; j++, bit++) {
+            if (val & (1UL << j)) {
+                host_mask[bit / host_bits] |= 1UL << (bit % host_bits);
```
Comment (powerjg) [inline src/arch/riscv/linux/se_workload.cc]: Please use the functions in `bitfield.hh` instead of manually reimplementing them. This will make bugs less likely.

File: src/arch/riscv/linux/se_workload.cc [hunk 3]
```diff
@@ -123,6 +125,179 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+              VPtr<> cpus)
+{
+    int ret, i, tmp;
+    size_t host_mask_size, target_mask_size;
+    unsigned long *host_mask;
+
+    tmp = (8 * sizeof(cpusetsize));
+    target_mask_size = ((cpusetsize + tmp - 1) / tmp) * sizeof(cpusetsize);
+    host_mask_size = (target_mask_size + (sizeof(*host_mask) - 1)) &
+                     ~(sizeof(*host_mask) - 1);
```
Comment (powerjg) [inline src/arch/riscv/linux/se_workload.cc]: Can you use functions in `bitfield.hh` to make this more readable?

If that's impossible, then adding comments to explain the logic is necessary.

File: src/arch/riscv/linux/se_workload.cc [hunk 4]
```diff
@@ -123,6 +125,179 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+template <class OS>
+static SyscallReturn
+riscvHWProbeFunc(SyscallDesc *desc, ThreadContext *tc,
+                 VPtr<> pairs,
+                 typename OS::size_t pair_count,
+                 typename OS::size_t cpusetsize,
+                 VPtr<> cpus,
+                 unsigned int flags)
+{
+    int ret;
```
Comment (powerjg) [inline src/arch/riscv/linux/se_workload.cc]: This should be on line 284

File: src/arch/riscv/linux/se_workload.cc [hunk 5]
```diff
@@ -123,6 +125,179 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+
+    /* flags must be 0 */
+    if (flags != 0) {
+        return -EINVAL;
+    }
+
+    /* check cpu set */
+    if (cpusetsize != 0) {
+        ret = cpu_set_valid<OS>(tc, cpusetsize, cpus);
```
Comment (powerjg) [inline src/arch/riscv/linux/se_workload.cc]:
```suggestion
        int ret = cpu_set_valid<OS>(tc, cpusetsize, cpus);
```

**Review by powerjg — DISMISSED** [review]
Thanks for the contribution! I would be happy to see support for hardware probing merged. The current code is pretty hard to follow and I believe it can be simplified by using some of gem5's built in functions/APIs. See below for details.

**Comment by 5265325** [comment]
@powerjg I rewrote the code. The code is now closer to the linux kernel than to qemu to improve readability.

**Comment by ivanaamit** [comment]
@5265325, Could you please mark the change requests as resolved if you've addressed them? This will make it easier for others to review or re-review. Thank you.

**Comment by 5265325** [comment]
> marked this conversation as resolved.

I have marked all conversations as resolved.  Is that right?

File: src/arch/riscv/linux/se_workload.cc [hunk 6]
```diff
@@ -371,6 +735,8 @@ SyscallDescTable<SEWorkload::SyscallABI64> EmuLinux::syscallDescs64 = {
     { 241,  "perf_event_open" },
     { 242,  "accept4" },
     { 243,  "recvmmsg" },
+    { 244,  "specific_syscall" },
```
Comment (hnpl) [inline src/arch/riscv/linux/se_workload.cc]: Can you remove this line? I don't think RISC-V has the system call 244.

https://github.com/torvalds/linux/blob/88264981f2082248e892a706b2c5004650faac54/scripts/syscall.tbl#L292

File: src/arch/riscv/linux/se_workload.cc [hunk 7]
```diff
@@ -710,6 +1076,8 @@ SyscallDescTable<SEWorkload::SyscallABI32> EmuLinux::syscallDescs32 = {
...
```
Comment (hnpl) [inline src/arch/riscv/linux/se_workload.cc]: Can you remove this line? I don't think RISC-V has the system call 244.

https://github.com/torvalds/linux/blob/88264981f2082248e892a706b2c5004650faac54/scripts/syscall.tbl#L292

File: src/arch/riscv/linux/se_workload.cc [hunk 8]
```diff
@@ -123,6 +125,368 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+    case RiscvLinux::BaseBehavior:
+        {
+            MISA misa = tc->readMiscRegNoEffect(MISCREG_ISA);
+            RiscvLinux::key_base_behavior_t *base_behavior = (RiscvLinux::key_base_behavior_t *)&pair->value;
+            if (misa.rvi && misa.rvm && misa.rva) {
+                base_behavior->ima = 1;
+            }
+        }
+        break;
+    case RiscvLinux::IMAExt0:
```
Comment (hnpl) [inline src/arch/riscv/linux/se_workload.cc:291]: You don't need to make any change here.

Though, it'll be great if we have a single entity managing all riscv extensions. Right now, we have two places that do this, here and https://github.com/gem5/gem5/blob/ba02266260a1babc6c05d44ee9f24c87b83d5fd7/src/arch/riscv/RiscvISA.py

File: src/arch/riscv/linux/se_workload.cc [hunk 9]
```diff
@@ -123,6 +125,368 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+    return pair->value == other_pair->value;
+}
+
+static inline int
+get_cpu_online_mask(ThreadContext *tc,
+                   uint64_t **cpu_online_mask,
+                   size_t *cpu_mask_size)
+{
+    /* 8-bytes up-boundary alignment */
+    *cpu_mask_size = (tc->getSystemPtr()->threads.size() + sizeof(**cpu_online_mask) - 1) /
```
Comment (hnpl) [inline src/arch/riscv/linux/se_workload.cc]: Note: this is rounding the `number_of_threads` to the next multiple of `size_of_each_mask`.

File: src/arch/riscv/linux/se_workload.cc [hunk 10]
```diff
@@ -123,6 +125,368 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
     return 0;
 }
 
+static inline void
+cpumask_set_cpu(unsigned int cpu,
+               uint64_t *dstp)
```
Comment (hnpl) [inline src/arch/riscv/linux/se_workload.cc]: style-nit: please fix the line break here and below. If the params fit within 79 characters, you don't need to break the line.

```cpp
cpumask_set_cpu(unsigned int cpu, uint64_t *dstp)
```

File: src/arch/riscv/linux/se_workload.cc [hunk 11]
```diff
@@ -123,6 +125,368 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+    case RiscvLinux::Cpuperf0:
+        return true;
+    }
+
+    return false;
+}
+
+static inline bool
+riscv_hwprobe_pair_cmp(RiscvLinux::riscv_hwprobe *pair,
+                      RiscvLinux::riscv_hwprobe *other_pair)
```
Comment (hnpl) [inline src/arch/riscv/linux/se_workload.cc]: style-nit: please fix the params alignment here and below.

```cpp
riscv_hwprobe_pair_cmp(RiscvLinux::riscv_hwprobe *pair,
                       RiscvLinux::riscv_hwprobe *other_pair)
```

**Review by hnpl — COMMENTED** [review]
I left a few comments, most are notes for future improvement, but there are a few style nits that need to be fixed.

The code logic looks okay to me as it follows the kernel implementation.

**Comment by hnpl** [comment]
It would be nice if the struct for the cpu_mask to simplify the code.

**Comment by 5265325** [comment]
> It would be nice if the struct for the cpu_mask to simplify the code.

I have fixed the style problem you mentioned. In addition, I added the cpumask_t struct to simplify the code.

**Review by hnpl — APPROVED** [review]
Thanks!

**Review by Harshil2107 — DISMISSED** [review]
Not an expert but it ltgm. I also looks like all the requested changes were addressed.

File: src/arch/riscv/linux/se_workload.cc [hunk 12]
```diff
@@ -134,6 +136,381 @@ unameFunc32(SyscallDesc *desc, ThreadContext *tc, VPtr<Linux::utsname> name)
...
+
+    int ret = 0;
+    RiscvLinux::riscv_hwprobe *pair;
+    RiscvLinux::cpumask_t *cpu_online_mask = get_cpu_online_mask(tc);
+    RiscvLinux::cpumask_t *cpus = cpumask_malloc(tc);
+    BufferArg pairs_buf(pairs, sizeof(RiscvLinux::riscv_hwprobe) * pair_count);
+
+    if (cpu_online_mask == nullptr || cpus == nullptr) {
+        ret = -ENOMEM;
+        goto exit;
```
Comment (powerjg) [inline src/arch/riscv/linux/se_workload.cc]: I would strongly prefer not having "goto" and instead duplicating the code. It's only used in one or two places, so I don't think that it's worth the downsides.

Also, I strongly prefer *always* using `{ }` around if/while statements. I have had to fix multiple very difficult to find bugs because people want to save 2 characters 😉

Comment (powerjg) [reply src/arch/riscv/linux/se_workload.cc]: Re: `goto`: There's only three places in the entire gem5 codebase this is used. Let's not add another, if we can possibly avoid it.

Comment (5265325) [reply src/arch/riscv/linux/se_workload.cc]: I've removed all `goto` and added `{ }`.

**Review by powerjg — APPROVED** [review]
